qiskit/providers/__init__.py

```diff
@@ -110,6 +109,391 @@
+reflect the latest version of this interface.
+
+Custom Basis Gates
+^^^^^^^^^^^^^^^^^^
+
+1. If your backend doesn't use gates in the Qiskit circuit library
+(:mod:`qiskit.circuit.library`) you can integrate support for this into your
+provider. The basic method for doing this is first to define a
+:class:`~qiskit.circuit.Gate` class for all the custom gates in the basis
+set. For example::
+
+    import numpy as np
+
+    from qiskit.circuit import Gate
+    from qiskit.circuit import QuantumCircuit
+
+    class SYGate(Gate):
+        def __init__(self, label=None):
+            super().__init__("sy", 1, [], label=label)
+
+        def _define(self):
+            qc = QuantumCircuit(1)
+            q.ry(np.pi / 2, 0)
+            self.definition = qc
+
+The key thing to ensure is that any custom gates in your Backend's basis set
+is that your custom gate's name attribute (the first param on
+``super().__init__()`` in the ``__init__`` defintion above) does not conflict
+with the name of any other gates. The name attribute is what is used to
+identify the gate in the basis set for the transpiler. If there is a conflict
+the transpiler will not know which gate to use.
+
+2. After you've defined the custom gates to use for the Backend's basis set
+then you need to add equivalence rules to the standard equivalence library
+so that the :func:`~qiskit.compiler.transpile` function and
+:mod:`~qiskit.transpiler` module can convert an arbitrary circuit using the
+custom basis set. This can be done by defining equivalent circuits, in terms
+of the custom gate, for standard gates. Typically if you can convert from a
+:class:`~qiskit.circuit.library.CXGate` (if your basis doesn't include a
+standard 2 qubit gate) and some commonly used single qubit rotation gates like
+the :class:`~qiskit.ciruit.library.HGate` and
+:class:`~qiskit.circuit.library.UGate` that should be sufficient for the
+transpiler to translate any circuit into the custom basis gates. But, the more
+equivalence rules that are defined from standard gates to your basis the more
+efficient translation from an arbitrary circuit to the target basis will be
+(although not always, and there is a diminishing margin of return).
+
+For example, if you were to add some rules for the above custom ``SYGate``
+we could define the :class:`~qiskit.circuit.library.U2Gate` and
+:class:`~qiskit.circuit.library.HGate`::
+
+    from qiskit.circuit.equivalence_library import SessionEquivalenceLibrary
+    from qiskit.circuit.library import HGate
+    from qiskit.circuit.library import ZGate
+    from qiskit.circuit.library import RZGate
+    from qiskit.circuit.library import U2Gate
+
+
+    # H => Z SY
+    q = qiskit.QuantumRegister(1, "q")
+    def_sy_h = qiskit.QuantumCircuit(q)
+    def_sy_h.append(ZGate(), [q[0]], [])
+    def_sy_h.append(SYGate(), [q[0]], [])
+    SessionEquivalenceLibrary.add_equivalence(
+        HGate(), def_sy_h)
```

Comment:
It might be worth mentioning that this can also be done through `HGate().add_decomposition(def_sy_h)` (though for sake of clarity, having the full example here makes sense).